### ISSUE TITLE
about uvcgan2

### ISSUE THREAD

**Issue by Aike2071** [LS4GAN/uvcgan2#44]
I achieved pretty good result using uvcgan on my dataset,but when I turn to uvcgan2,the result is confusing.

![Image](https://github.com/user-attachments/assets/bce2efd6-6b09-4ecc-9aed-a85b25dede8d)

![Image](https://github.com/user-attachments/assets/e0fc4e6f-614c-4b10-a44f-89622af1a54a)

as you can see on the image,uvcgan2 just adds a filter to original image instead of performing image translation.I didn't make any changes on the code except for dataroot.What could be the possible reason for this?Your reply is appreciated.

**Comment by Aike2071** [comment]
here is what uvcgan can achieve

![Image](https://github.com/user-attachments/assets/48e33622-0216-4169-b215-2201f8758127)

**Comment by usert5432** [comment]
Hi @Aike2071, that is an interesting behavior.

We can do a few things to debug what is happening. First, I think, the most likely reason is the loss saturation. To see if that is the case, could you check the value of the training losses? Does disc_a or disc_b reach values close to 0 or 1? If so, the training may have collapsed. The solution will depend on when exactly the collapse happens.

If the losses seem fine, we may need to investigate other possibilities.

**Comment by Aike2071** [comment]
Hi @usert5432 ,
Here are my loss curves and training procedures.First, I pre-trained the model and got such ideal results.Then I started training with the checkpoint of epoch 1800 as the pre-trained model.Unfortunately, I guess the collapse happened at the very beginning?

![Image](https://github.com/user-attachments/assets/fb7dc151-0698-4e62-8ea1-8e25d521f87f)
![Image](https://github.com/user-attachments/assets/1d28bd11-4472-4cb5-8cfc-9d8045494ea6)

**Comment by Aike2071** [comment]
And one thing I've just noticed is that there is no consist image generated.Is that normal?

**Comment by usert5432** [comment]
Hi @Aike2071,

Thank you for sharing the losses. Let me provide my thoughts about them:
1. The generator losses gen_ab and gen_ba look quite normal.
2. The discriminator losses disc_a and disc_b look quite normal as well. This is very strange, as the discriminator should easily detect that the images do not look realistic. If the discriminator detected that -- its loss should become 0.
3. Cycle consistency loss may be a bit large, but in the normal range. Could you verify if the `reco_a` image matches `real_a` image?
4. Identity losses are good, which means the generator is not broken.

The only thing that comes to my mind, that can cause this behavior, is EMA averaging of the generators: for training we use a normal generator, but for inference an EMA averaged copy. With the default configuration, it takes about ~10k training iterations to update the value of the copy. If you dataset happens to be too small, then the training may be simply too short for the copy to be properly updated and we observe partial results.

Another possibility (less likely) is that the discriminator is completely broken. But, nothing immediate comes to my mind about what could be wrong there.

Would it be possible for you to share a training configuration you used and the size of datasets you have? Perhaps, some clues may be there.

> And one thing I've just noticed is that there is no consist image generated.Is that normal?

Yes, by default, the consistency loss is disabled, so consistency images are not generated.

**Comment by Aike2071** [comment]
Thank you for sharing your thoughts!
I checked the reco_a and reco_b,all the images are like this one here.

![Image](https://github.com/user-attachments/assets/a86bd82f-5631-415a-9f62-f16464066924)

and here is my training configuration.I have 1758 images for trainA and 1403 for trainB.I don't know if this is too small for GANs because I'm new on this field.
``
{
    "batch_size": 32,
    "data": {
        "datasets": [
            {
                "dataset": {
                    "domain": "a",
                    "name": "cyclegan",
                    "path": "/home/aike/Desktop/uvcgan/scripts/data/lunar_A2B"
                },
                "shape": [
                    3,
                    256,
                    256
                ],
                "transform_test": null,
                "transform_train": [
                    {
                        "name": "resize",
                        "size": 286
                    },
                    {
                        "name": "random-crop",
                        "size": 256
                    },
                    "random-flip-horizontal"
                ]
            },
            {
                "dataset": {
                    "domain": "b",
                    "name": "cyclegan",
                    "path": "/home/aike/Desktop/uvcgan/scripts/data/lunar_A2B"
                },
                "shape": [
                    3,
                    256,
                    256
                ],
                "transform_test": null,
                "transform_train": [
                    {
                        "name": "resize",
                        "size": 286
                    },
                    {
                        "name": "random-crop",
                        "size": 256
                    },
                    "random-flip-horizontal"
                ]
            }
        ],
        "merge_type": "unpaired",
        "workers": 10
    },
    "discriminator": {
        "model": "basic",
        "model_args": {
            "shrink_output": false
        },
        "optimizer": {
            "betas": [
                0.5,
                0.99
            ],
            "lr": 0.0001,
            "name": "Adam"
        },
        "spectr_norm": true,
        "weight_init": {
            "init_gain": 0.02,
            "name": "normal"
        }
    },
    "epochs": 500,
    "generator": {
        "model": "vit-modnet",
        "model_args": {
            "activ": "gelu",
            "activ_output": "sigmoid",
            "embed_features": 384,
            "features": 384,
            "ffn_features": 1536,
            "modnet_activ": "leakyrelu",
            "modnet_demod": true,
            "modnet_downsample": "conv",
            "modnet_features_list": [
                48,
                96,
                192,
                384
            ],
            "modnet_norm": null,
            "modnet_rezero": false,
            "modnet_upsample": "upsample-conv",
            "n_blocks": 12,
            "n_ext": 1,
            "n_heads": 6,
            "norm": "layer",
            "rezero": true,
            "style_bias": true,
            "style_rezero": true
        },
        "optimizer": {
            "betas": [
                0.5,
                0.99
            ],
            "lr": 5e-05,
            "name": "Adam"
        },
        "spectr_norm": false,
        "weight_init": {
            "init_gain": 0.02,
            "name": "normal"
        }
    },
    "gradient_penalty": {
        "center": 0,
        "lambda_gp": 0.01,
        "mix_type": "real-fake",
        "reduction": "mean"
    },
    "loss": "lsgan",
    "model": "uvcgan2",
    "model_args": {
        "avg_momentum": 0.9999,
        "head_config": {
            "activ": "leakyrelu",
            "input_features": 512,
            "name": "batch-norm-2d",
            "output_features": 1
        },
        "head_queue_size": 3,
        "lambda_a": 10.0,
        "lambda_b": 10.0,
        "lambda_idt": 0.5
    },
    "scheduler": null,
    "seed": 0,
    "steps_per_epoch": 2000,
    "transfer": {
        "allow_partial": false,
        "base_model": "lunar_pre/model_m(autoencoder)_d(None)_g(vit-modnet)_pretrain-uvcgan2",
        "fuzzy": null,
        "strict": true,
        "transfer_map": {
            "gen_ab": "encoder",
            "gen_ba": "encoder"
        }
    }
}
 ``

**Comment by usert5432** [comment]
Hi @Aike2071,

I think, there is a plausible case for the EMA model not being updated fully:
1. In the config file, I see `"batch_size": 32`.
2. The training dataset has size of 1403, which translates to ~44 training iterations per epoch.
3. In 500 epochs, the EMA network will be updated just about 20k times, which is barely enough to produce something non-random. (For reference, in the original training, the EMA network is updated about 1,000k times)

Would it be possible for you to try the training with a batch size == 1? Alternatively, you could try reducing the EMA averaging momentum from `"avg_momentum": 0.9999` down to say 0.999 or even lower.

**Comment by Aike2071** [comment]
Hi @usert5432 ,
I thought larger batch_size leads to faster training, but I wasn’t aware of its relationship with EMA averaging.
I will try to train again with the suggestions you provided.

**Comment by Aike2071** [comment]
There is one other thing that confuses me.When I start train translation,the training proedure always got frozen `somehow.The GPU and CPU usage drop to 0 but there are still RAM and VRAM usage.No any error message.
``
done set_input 
done optimization steps 
done metrics.update 
done set_postfix 
Epoch 9 / 500:   1%|       | 6/701 [00:05<08:14,  1.41it/s, gen_ab=0.474, gen_ba=0.31, cycle_a=0.148, cycle_b=0.181, disc_a=0.251, disc_b=0.153, idt_a=0.0194, idt_b=0.034, gp_a=0.00017, gp_b=0.0018]
done progbar.update 
``
I added some printed message to find out where exactly the freeze begins.The last output message was from training_epoch().
Similar thing occurred before and I solved it by using another device.Have you ever encountered something like this before?

**Comment by usert5432** [comment]
>  Similar thing occurred before and I solved it by using another device.Have you ever encountered something like this before?

Hmm. I only saw a similar situation in 3 cases:
1. when training on a machine with multiple GPUs. Somehow, pytorch freezes for an extended period of time there. Solution is to bind the training to a single GPU.
2. when batch size is large, but the number of workers is small or disk is slow. Solution is to either increase the number of workers or reduce batch size (or move data to a faster disk if it is a problem).
3. there is also an occasional random freezing because of a bug in a pytorch data loader. But, in this case it freezes forever and one needs to restart the training.

**Comment by Aike2071** [comment]
I guess it’s case 3 I am encountering. The freezing just happens,even though I reduced batch size.
Anyway, thank you very much for your advice and help!
